### PR TITLE
Add support for Llama3-70b

### PR DESCRIPTION
**Pull request by bhavya01**
Test run output: https://gist.github.com/bhavya01/07dd88d76f3d339de664ebecc3dc035a

llama3 shards the embeddings differently than llama2. So, I created a new default_sharding file for it.

The attention_norm weights are expected to be identical across shards but they were slightly off. So, increased the tolerance while converting checkpoints.

### CODE REVIEW

File: convert_checkpoints.py [hunk 1]
```diff
@@ -179,17 +179,19 @@ def _merge_llama_weights(
...
       with torch.no_grad():
         if kind in ("ParallelEmbedding", "RowParallelLinear"):
           state_dict_for_key[key] = torch.cat(tensors, 1)
-        elif kind == "ColumnParallelLinear":
+        elif kind in ("ColumnParallelLinear", "VocabParallelEmbedding"):
           state_dict_for_key[key] = torch.cat(tensors, 0)
         else:
           if not all(
-              torch.allclose(tensors[0], tensor, atol=1e-6)
+              torch.allclose(tensors[0], tensor, atol=1e-2)
```
Comment (FanhaiLu1) [inline convert_checkpoints.py:194]: Any reason to loose condition by e**4 magnitude?

File: jetstream_pt/third_party/llama/model_exportable.py [hunk 2]
```diff
@@ -225,13 +225,12 @@ def get_quantized_embedding_weight_to_scaler_map():
     }
 
   @staticmethod
-  def get_weight_sharding_type():
+  def get_weight_sharding_type(model_name: str = "llama-3"):
```
Comment (FanhaiLu1) [inline jetstream_pt/third_party/llama/model_exportable.py:228]: Some potential issue with current logic: 

1: User might think llama-2 is default 
2: If user pass wrong model, we still return a sharding_dict

User may forgot give model_name or give a wrong model_name. I feel it's better to set default as None (or not set default value) and enforce user pass right model name. Have validation in this function and throw error if user didn't pass right model.

File: default_shardings/llama-3.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+
+# Sharding config for llama-3
+# Sharding should either be an int between 0 and rank - 1
+# signifying the axis to shard or -1 / null signifying replicated
+
+
+freqs_cis : -1 #  torch.complex64 (2048, 64)
+tok_embeddings.weight : 0 #  torch.float32 (vocab_size, 4096)
```
Comment (lsy323) [inline default_shardings/llama-3.yaml:8]: The sharding file seems to be the same as llama-2. What's the difference between the llama-2 and llama-3 sharding file?

From the change in `convert_checkpoints.py`, it seems that llama-3 weight is sharded in a different way. This sharding file is only used for model sharding during runtime. 

If this is the case, we don't need to have another sharding yaml file.

**Comment by FanhaiLu1** [comment]
> Test run output: https://gist.github.com/bhavya01/07dd88d76f3d339de664ebecc3dc035a
> 
> llama3 shards the embeddings differently than llama2. So, I created a new default_sharding file for it.
> 
> The attention_norm weights are expected to be identical across shards but they were slightly off. So, increased the tolerance while converting checkpoints.

The output of Llama3-70B dropped if we compare it with LLama2-7B. Can you create a bug to track it? 

There are repeated output in example: 
---- All output text.
 to give life a meaning. -Paul Thoreau
I believe the meaning of life is to give life a meaning. -Paul Thoreau
I believe the meaning of life is to give life a meaning. -Paul Thoreau
I believe the meaning of life is to give life a meaning. -Paul Thoreau

File: default_shardings/llama-3.yaml (hunk 3 above)
Comment (bhavya01) [reply default_shardings/llama-3.yaml:8]: The tok_embeddings.weight is sharded differently between llama-2 and llama-3. For llama-2, embeddings are sharded along axis 1 and for llama-3, they are sharded along axis 0. But I agree, that it shouldn't make a difference in accuracy during runtime. If you think that it is better to keep the same sharding for both of them then I can revert this change.

File: convert_checkpoints.py (hunk 1 above)
Comment (bhavya01) [reply convert_checkpoints.py:194]: The layer norm weights in llama-3 are not consistent across shards. I don't know why is this the case. These weights are expected to be replicated. It errors out if we don't reduce the precision here.

**Comment by bhavya01** [comment]
> > Test run output: https://gist.github.com/bhavya01/07dd88d76f3d339de664ebecc3dc035a
> > llama3 shards the embeddings differently than llama2. So, I created a new default_sharding file for it.
> > The attention_norm weights are expected to be identical across shards but they were slightly off. So, increased the tolerance while converting checkpoints.
> 
> The output of Llama3-70B dropped if we compare it with LLama2-7B. Can you create a bug to track it?
> 
> There are repeated output in example: ---- All output text. to give life a meaning. -Paul Thoreau I believe the meaning of life is to give life a meaning. -Paul Thoreau I believe the meaning of life is to give life a meaning. -Paul Thoreau I believe the meaning of life is to give life a meaning. -Paul Thoreau

Sorry, can you explain the problem a little bit more. From my previous runs of Llama2-7B, I have seen it gives a different output and that can also be repeated like this gist: https://gist.github.com/bhavya01/40a344e671a2e5dde980f163141545db

**Comment by FanhaiLu1** [comment]
> > > Test run output: https://gist.github.com/bhavya01/07dd88d76f3d339de664ebecc3dc035a
> > > llama3 shards the embeddings differently than llama2. So, I created a new default_sharding file for it.
> > > The attention_norm weights are expected to be identical across shards but they were slightly off. So, increased the tolerance while converting checkpoints.
> > 
> > 
> > The output of Llama3-70B dropped if we compare it with LLama2-7B. Can you create a bug to track it?
> > There are repeated output in example: ---- All output text. to give life a meaning. -Paul Thoreau I believe the meaning of life is to give life a meaning. -Paul Thoreau I believe the meaning of life is to give life a meaning. -Paul Thoreau I believe the meaning of life is to give life a meaning. -Paul Thoreau
> 
> Sorry, can you explain the problem a little bit more. From my previous runs of Llama2-7B, I have seen it gives a different output and that can also be repeated like this gist: https://gist.github.com/bhavya01/40a344e671a2e5dde980f163141545db

I see, looks like there are accuracy issues in quantization. When I mentioned quality drop, I compared it with bfp16. For quantization accuracy issue, it's not related with this cl.

File: convert_checkpoints.py (hunk 1 above)
Comment (FanhaiLu1) [reply convert_checkpoints.py:194]: @qihqi are you ok with 1e-2 gap? I feel it's risky when we loose condition by e**4 magnitude for a single tensor.

Comment (qihqi) [reply convert_checkpoints.py:194]: yeah that is fine

File: default_shardings/llama-3.yaml (hunk 3 above)
Comment (qihqi) [reply default_shardings/llama-3.yaml:8]: they shouldnt be sharded differently -- the only difference would be performance; lets run with both and keep the faster one.

**Review by qihqi — APPROVED** [review]

**Review by FanhaiLu1 — APPROVED** [review]
